open-craft/ecommerce
# [BB-3801] Implement pending order status handling

This PR implements the changes for handling the pending order status.

**Testing instructions**:
* The setup instructions are similar to those mentioned in https://github.com/open-craft/ecommerce/pull/21.
* Add 2 additional configuration parameters - `encryption_key` (a randomly generated string of a reasonable length), `salt` (a randomly generated string of a reasonable length) to the existing payment processor configuration.
* Go through the typical successful payment flow and verify that it works okay and there are no issues introduced by this PR.
* In the HyperPay simulator that shows up after entering the card details, select the status as `pending`. Verify that the user is taken to a separate page that shows a message about the transaction being in progress. The page should refresh once every 30 seconds unless configured otherwise.
* In a non-test environment, the status will eventually change to either success or failure, but it doesn't in this case. So just navigate back to the LMS to try other tests.
* Verify that the payment flow where you choose an error result also works okay without any issues.
* Try any other scenario to try and break this code and find bugs :)

**Author notes and concerns**:
* HyperPay has a lot of sub-categories of result codes ([docs](https://hyperpay.docs.oppwa.com/reference/resultCodes)) for each major category - success, pending, rejected. This PR assumes that 
  * "Successfully processed transactions that should be manually reviewed" are successful since there is no automated way for e-commerce to do a manual review. I don't think there is any way to notify the admins about it either unless we want to write something new.
  * The pending transactions of the type "where the status of a transaction can change even after several days" are treated as failures.
* The main view handling the response from HyperPay redirects to a separate page with the relevant information encrypted in the URL parameter, for the pending status. This is done to obfuscate it from the user like it is typically done in payment flows.
